### ISSUE TITLE
version conflicts?

### ISSUE THREAD

**Issue by max-mapper** [mikeal/level-sleep#1]
right now when doing `levelsleep.pull` it just gets the remote sleep feed and calls `levelsleep.put(key, val)` for each entry

the way that `put` is implemented means that if you call `pull` once you'll end up with data like this (formatted for readability):

```
{ key: 'foo', value: { val: 'hello', _seq: 1, _id: 'foo' } }
{ key: 'bar', value: { val: 'hello', _seq: 2, _id: 'bar' } }
{ key: '1', value: { _id: 'foo', _seq: 1 } }
{ key: '2', value: { _id: 'bar', _seq: 2 } }
```

but then calling again (without since) results in this

```
{ key: 'foo', value: { val: 'hello', _seq: 3, _id: 'foo' } }
{ key: 'bar', value: { val: 'hello', _seq: 4, _id: 'bar' } }
{ key: '1', value: { _id: 'foo', _seq: 1 } }
{ key: '2', value: { _id: 'bar', _seq: 2 } }
{ key: '3', value: { _id: 'foo', _seq: 3 } }
{ key: '4', value: { _id: 'bar', _seq: 4 } }
```

is your assumption that compaction should clean up the duplicate sequences or do you think we need some sort of revision system so that we can prevent the duplicates in the first place?

**Comment by max-mapper** [comment]
theres also a tangentially related discussion here https://github.com/maxogden/dat/issues/10

**Comment by mikeal** [comment]
Actually, you end  up with:

```
{ key: 'foo', value: { val: 'hello', _seq: 1, _id: 'foo' } }
{ key: 'bar', value: { val: 'hello', _seq: 2, _id: 'bar' } }
{ key: 'foo', value: { val: 'hello', _seq: 3, _id: 'foo' } }
{ key: 'bar', value: { val: 'hello', _seq: 4, _id: 'bar' } }
{ key: '1', value: { _id: 'foo', _seq: 1 } }
{ key: '2', value: { _id: 'bar', _seq: 2 } }
{ key: '3', value: { _id: 'foo', _seq: 3 } }
{ key: '4', value: { _id: 'bar', _seq: 4 } }
```

Because we also don't remove the old documents until compaction which is why `get()` does a `peekLast` for the "current" document.

i talked with @jchris about this and the "sparse sequence index" CouchDB uses is just an optimization.

By leaving the old sequences (and documents) until compaction we get much faster write times (we don't need to read before write, ever) and we have the ability to play back and even retrieve the old documents/sequences.

I don't know which optimization is more useful at this point in time but I do plan on adding an `add(key, value)` method to the database that will only write the key if it is not already there simply because I need it in my own work if i want to stop running compaction **a lot**.

**Comment by mikeal** [comment]
oh, also notice that `include_data` will actually return the `data` for each sequence, not the _current_, so the SLEEP feed is the full append log.

**Comment by jchris** [comment]
If you are doing writes without read, then you can't check for matching
revision identifiers. This means you can't do proper conflict detection and
management. Just saying.

On Wednesday, August 21, 2013, Mikeal Rogers wrote:

> oh, also notice that include_data will actually return the data for each
> sequence, not the _current_, so the SLEEP feed is the full append log.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mikeal/level-sleep/issues/1#issuecomment-23051403
> .

## 

Chris Anderson
http://jchrisa.net
http://www.couchbase.com

**Comment by max-mapper** [comment]
it seems like we could have a `level-conflicts` module that provided the conflict layer, but the default should be to sacrifice disk space in the name of write throughput (the simplest option)

**Comment by mikeal** [comment]
@maxogden what do you consider a _conflict_ being that there are no revisions?

**Comment by mikeal** [comment]
@jchris yup, exactly :) that's why this is `level-sleep` and not `couchup`. 

I do the revision checking in `couchup` but specifically don't do anything with revisions in `level-sleep`. For some of the data collection I've done lately the _no revs_ approach is actually preferred and I've been using this module a lot for that.

**Comment by max-mapper** [comment]
@mikeal i'm just saying the concept of revisions and conflicts could be 100% layered on top of level-sleep, which would just do the dumbest possible thing

also shouldnt `.pull` include `since` by default?

**Comment by mikeal** [comment]
it is, you just need to pass the `save` option to signal that it should cache the last write sequence for the remote name.

https://github.com/mikeal/level-sleep/blob/master/index.js#L196

**Comment by mikeal** [comment]
@maxogden so, a little about revs and conflicts.

I think if you want to write a revision system, and get it to replicate, and get it to respond to conflicts in a predictable way, it should just be written as an extension of the SLEEP protocol.

For instance, the way `couchup` does revisions is different enough from `CouchDB` that replicating back and forth will cause unnecessary revision conflicts on the CouchDB side.  I could easily write a spec for how `couchup` does revisions (linear revisioning) as an extension of SLEEP and people could replicate all over the place like **w0ah!** But, the important thing is that anyone who _doesn't_ do revisioning can still get the SLEEP feed from `couchup` with the revision information and just ignore it and still be able to pull down the full data set.
